Allows naming signals

JobJob opened the pull request:
Useful for debugging and could be used for visualising signal graphs
They will get an auto-incremented node-number as their name if none is provided.

Aside: I'm a little busy right now so if you want to make changes yourself that would be so lovely :D

JobJob commented:
Oh, looks like I only implemented this for Signal and map... Don't really have time to do the others right now 😬

JobJob commented:
Doing a bit of cleanup. Tests are failing.

shashi commented on src/core.jl:
```diff
@@ -9,6 +9,9 @@ const debug_memory = false # Set this to true to debug gc of nodes
 const io_lock = ReentrantLock()
+const node_count = Ref(1) #1 since auto_name gets called before count is incremented
+
+auto_name() = string(node_count[])
```
maybe `string("signal-", node_count[])`?

shashi replied:
increment should also be done here.

JobJob commented:
Ok updated and (hastily) added the name option to most operators.

N.b. API for `onerror` callback is changed node where error occurred also passed to the callback now.

shashi commented on src/core.jl:
```diff
@@ -17,6 +20,11 @@ if !debug_memory
+        function Signal(v, parents, actions, alive, pres, name)
+            node_count[] += 1
+            n=new(v,parents,actions,alive,pres,name)
+        end
```
Can remove this if increment is moved to `auto_name`

JobJob replied:
yep - done

shashi commented on src/operators.jl:
```diff
@@ -235,8 +236,8 @@ Flatten a signal of signals into a signal which holds the
-function flatten(input::Signal; typ=Any)
-    n = Signal(typ, value(value(input)), (input,))
+function flatten(input::Signal; typ=Any, name=auto_name())
+    n = Signal(typ, value(value(input)), (input,); name=name)
```
I wonder if instead of adding keyword arguments to everything we can define `name!(signal, "name")` to set the name for signals you want to name.

It might also be nicer to default have default names such as "map(signal-1, signal-2)-1", "filter(signal-1)-2" etc.

JobJob replied:
Great ideas, but I've no time right now. 
Add them later?

JobJob replied:
I think naming them on creation makes sense, and the kw args are already there now. I've also added a `rename!(s::Signal, name::String)`

shashi commented:
@JobJob Thanks! no particular hurry here. I'll have a look at this later and make the changes.

JobJob commented:
Alright great, please do, thanks.

JobJob commented:
Added a new commit, I think it's probably ready.

Dug this out from a comment on an old commit, since it's the one remaining discussion point:
> I wonder if instead of adding keyword arguments to everything we can define name!(signal, "name") to set the name for signals you want to name.

Auto-naming them on creation makes sense to me, and the kw args are already there now. Any good reasons to change it to explicit `name!`? What would the default be in that case - empty string?

Also I've added a `rename!(s::Signal, name::String)`

> It might also be nicer to default have default names such as "map(signal-1, signal-2)-1", "filter(signal-1)-2" etc.

This is implemented now.

shashi commented on src/core.jl:
```diff
@@ -17,6 +17,10 @@ if !debug_memory
         preservers::Dict
+        name::String
+        function Signal(v, parents, actions, alive, pres, name)
+            n=new(v,parents,actions,alive,pres,name)
```
what's the use of this constructor?

JobJob replied:
No use, maybe it's vestigial from an earlier version where I had autoname in there? can't remember, either way, let's remove it

shashi commented:
Sweet!

> Any good reasons to change it to explicit `name!`?

Just to avoid having an extra kwarg to maintain for every function. But I'm not crazy about it though. This looks good!

shashi commented on src/operators.jl:
```diff
@@ -24,9 +24,10 @@ export map,
 Transform signal `s` by applying `f` to each element. For multiple signal arguments, apply `f` elementwise.
 """
 function map(f, input::Signal, inputsrest::Signal...;
```
I think because it was causing ambiguity for the `map(f)` method sinc Base defines `map(f, xs...)` too.

JobJob replied:
ok great, thanks.

JobJob commented:
I can't merge this btw, don't have [the power](https://www.youtube.com/watch?v=V8h8snfYidg). I will squash the commits.

JobJob commented:
Squashed.

shashi commented:
huh that's weird what's the point of an org member then. I've invited you as a collaborator on this repo.

tknopp commented:
@shashi: I think you can selectively control the repositories for an org member (with the default of no access) whereas a collaborator has access to every repo

timholy commented on src/core.jl:
```diff
@@ -223,7 +244,7 @@ function run(n::Int=typemax(Int))
                 else
                     bt = catch_backtrace()
-                    msgval.onerror(msgval.node, msgval.value, CapturedException(err, bt))
+                    msgval.onerror(msgval.node, msgval.value, node, CapturedException(err, bt))
```
The docstring for `push!` needs to be updated to specify the new arguments for `onerror`. This is a breaking change (the next tag will have to be 0.4.0), as anyone using `onerror` now needs to change their callback functions. Unfortunately, I don't see how one can emit a deprecation warning, except possibly to look for this in failures of the `try`. It also needs to be advertised somehow, e.g., in a `NEWS.md` file.

tkelman commented on src/operators.jl:
```diff
@@ -143,9 +145,9 @@ Merge many signals into one. Returns a signal which updates when
 the value of the *youngest* input signal is taken.
 """
-function merge(inputs...)
+function merge(inputs...; name=auto_name!("merge", inputs...))
```
is this missing argument types? can this avoid pirating Any... inputs?
